BIP39 Implementation

evanlinjin opened the pull request:
### Description

This is a continuation of PR #607 which closes #561

This PR includes commits for own implementation of `PBKFD2`.
I've also modified the `.gitignore`, I hope that is okay.

### Notes to the reviewers

Although complete, I still have some security concerns for the current implementation ([please check my comment below](https://github.com/bitcoindevkit/bdk/pull/644#issuecomment-1172943386)).

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] Add `pbkfd2` implementation
* [x] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

atalw committed:
MOVEONLY: Move bip39.rs to bip39/mod.rs

atalw committed:
BIP39: Custom implementation

There are a couple of features that have been implemented in this
commit:
- Parse mnemonic string to Mnenomic type
- Generate Mnemonic from entropy
- Derive seed from Mnemonic (with and without passphrase)
- All language wordlists (with verification test to ensure they were
  untampered)
- Mnemonic test vectors from BIP39
- Error handling

Function names have mostly been kept the same to maintain backwards
compatability.

Co-authored-by: Vladimir Fomene <vladimirfomene@gmail.com>

atalw committed:
BIP39: Add feature flags for mnemonic languages

vladimirfomene committed:
BIP39: Verify checksum logic and new tests

* `Mnemonic::parse_in` now verifies the checksum against the entropy.

* Add test: Make sure `from_entropy_in` produces error if length of
  entropy bits is less than 128, greater than 256 and not a multipe of 32.

* Add test: invalid mnemonic sentence.Throw error if mnemonic sentence
  is less than 12 words or greater than 24 words or number of words is
  not a multiple of six or the contains a word not in wordlist or has
  invalid checksum.

evanlinjin committed:
.gitignore for vscode

evanlinjin committed:
BIP39: Add own PBKFD2 implementation

Also removed unused dependencies

danielabrozzoni commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 - New MSRV set to `1.56.1`
 - Fee sniping discouraging through nLockTime - if the user specifies a `current_height`, we use that as a nlocktime, otherwise we use the last sync height (or 0 if we never synced)
+- Changed: Introduce own implementation of `bip39`.
```
nit: remove the initial `Changed:`

evanlinjin replied:
Resolved in b0033915e379cd365d2c1953b7ed6c6aee76679e

danielabrozzoni commented on src/keys/bip39/mod.rs:
```diff
@@ -37,9 +88,149 @@ pub enum WordCount {
+                None => return Err(Error::InvalidWord(word.to_string())),
+            };
+
+            indices.push(idx);
```
@vladimirfomene was saying, in #607:

> After calculating the indices for every word, you may want to recover the entropy, then calculate the checksum to compare it against the binary representation of mnemonic. In case, the calculated checksum does not match the checksum in the binary representation of the mnemonic, you will want to throw an error.


Looks like it's still a todo, so I'm posting it here

evanlinjin replied:
Resolved in b0033915e379cd365d2c1953b7ed6c6aee76679e

danielabrozzoni commented on src/keys/bip39/mod.rs:
```diff
@@ -37,9 +88,149 @@ pub enum WordCount {
+    }
+
+    /// Convert a mnemonic to a seed with an optional passphrase
+    fn to_seed(&self, passphrase: Option<String>) -> Seed {
```
From #607:

@vladimirfomene: It might not be a good idea to change the type of passphrase from &str to option<string> as that has the potential of breaking code which consumes this method.

@atalw: It makes sense to have the passphrase as Option as it really is optional, so if a breaking change is okay we can go ahead with this.

-----------

Personally, I agree that we should try not to break the API, and leave `P: Into<Cow<'a, str>` (https://docs.rs/bip39/latest/src/bip39/lib.rs.html#479-486) here

evanlinjin replied:
In my opinion, having a `P: Into<Cow<'a, str>>` makes little sense based on our implementation. I propose just using a `&str`, this way for most people, the API shouldn't break.

evanlinjin replied:
Changed in dd42594c78455d2e71c8cae86746ce9c0a2031c5. Let me know if it is sufficient!

https://github.com/bitcoindevkit/bdk/blob/dd42594c78455d2e71c8cae86746ce9c0a2031c5/src/keys/bip39/mod.rs#L204-L208

danielabrozzoni reviewed (commented):
Copy-pasted some review comments from #607 that it seems still need to be addressed

evanlinjin committed:
BIP39: Performance and readability changes

* Introduce `Language::word_map` method for faster word index finding.
* Readability changes to various `Mnemonic` methods and tests.
* Re-introduce various methods back into `Mnemonic`.
* `bip39::Error` no longer includes sensitive information.

evanlinjin committed:
BIP39: Various fixes and new tests

Fixes:
* Fixed implementations of `GeneratableKey` to work for all word lengths
* Fixed example in `rpcwallet`

Changes:
* Added various `derive`s for bip39 structures
* Added `Mnemonic::with_passphrase` method
* Added `TryFrom<uszie>` implementation for `WordCount`
* Introduced `Bip39TestVector` struct for more comprehensive testing
* Various refactoring

CI/CC Changes:
* Added `all-languages` feature to `[package.metadata.doc.rs]`
* Added `all-languages` feature to code coverage and CI tests

evanlinjin commented:
I still have some security concerns regarding the current implementation (although, I am no security expert, just based on what I've read on the internet). I will list them here, and hopefully someone knowledgeable enough could provide some clarity.

1. Should we avoid using heap memory, and keep everything on the stack? Apparently, heap memory is more prone to exploints. Reference: [github.com/shellphish/how2heap](https://github.com/shellphish/how2heap)

2. Is implementing `std::fmt::Display` and `Debug` a good idea? As it may leak the secrets to logs. We can potentially remove these implementations completely, or provide implementations with redacted secrets, or use an a crate such as [`secrey`](https://docs.rs/secrecy/latest/secrecy/index.html).

Thank you all in advance!

P.S. Test [`blockchain::esplora::bdk_blockchain_tests::test_sync_stop_gap_20`](https://github.com/bitcoindevkit/bdk/runs/7163681328?check_suite_focus=true) seems to fail occasionally.

danielabrozzoni commented:
> Should we avoid using heap memory, and keep everything on the stack? Apparently, heap memory is more prone to exploints. Reference: [github.com/shellphish/how2heap](https://github.com/shellphish/how2heap)

We could look into that, but it's better if we do so in a new PR. This one is already quite big, and the bigger it gets, the more difficult it is to collect reviews :)

> Is implementing std::fmt::Display and Debug a good idea? As it may leak the secrets to logs. We can potentially remove these implementations completely, or provide implementations with redacted secrets, or use an a crate such as [secrey](https://docs.rs/secrecy/latest/secrecy/index.html).

This, instead, I think should be tackled here: for now, avoiding Debug and Display (or manually implementing a really generic one) should be enough (with an appropriate comment on why we do so). I'd avoid adding YA dependency :)

evanlinjin commented:
Another aspect I've been thinking about, is the great majority of the time people will be using English (which shouldn't require Unicode normalization). For the passphrase, we can do a check only (and fail if not normalized).

Since normalization sometimes requires resizing the vector (so it's a heap operation), and for most people, it also means one less dependably.

evanlinjin committed:
BIP39: Remove potentially dangerous trait impls

`Mnemonic` contains sensitive data so we should ensure internal fields
are not easily leaked.

* Explicitly implement `fmt::Debug` and redact all fields.
* Explicitly implement `ToString` instead of `Display`.
* Remove various comparative `derive()`s.

evanlinjin commented:
> This, instead, I think should be tackled here: for now, avoiding Debug and Display (or manually implementing a really generic one) should be enough (with an appropriate comment on why we do so). I'd avoid adding YA dependency :)

Addressed in 08e1cc9ebfdb5f9e0f4bfd929eb71b94eedea89c.

afilini commented:
> Should we avoid using heap memory, and keep everything on the stack?

One advantage of this (on top of the extra safety) is that it would be much easier to then port to embedded hardware. We have many features in bdk which I guess are not really fit for hardware wallets, but mnemonics are for sure one that we'll need to have.

Are you able to do a rough estimation of how much longer/how much harder it would be to implement in this way?

afilini commented on src/keys/bip39/pbkdf2.rs:
```diff
@@ -0,0 +1,87 @@
+/// Password is the UTF8-NFKD-normalized result of mnemonic words separated by space.
+fn make_password<'a, W>(words: W) -> String
+where
+    W: Iterator<Item = &'a str>,
```
You could define `Item` as another generic that implements `AsRef<str>`. This should allow you to pass a vec of strings as well if you want

afilini commented on src/keys/bip39/pbkdf2.rs:
```diff
@@ -0,0 +1,87 @@
+
+/// Salt is the UTF8-NFKD-normalized result of (SALT_PREFIX + passphrase).
+fn make_salt(passphrase: &str) -> Cow<'static, str> {
+    let mut salt = Cow::from(SALT_PREFIX);
```
I understand the small performance benefit of reusing the ref as-is, but I don't think it's worth using `Cow` here especially considering that moving forward we'd like to avoid using the heap (even if we don't manage to finalize that transition in this PR)

afilini commented on src/keys/bip39/pbkdf2.rs:
```diff
@@ -0,0 +1,87 @@
+/// Make hmac-sha512 engine from password.
+/// The hmac engine is used as the pseudo-random function.
+fn make_prf(password: &str) -> HmacPRF {
+    HmacEngine::new(password.as_bytes())
```
I was gonna comment here that for extra safety we should re-normalize the string, but then I realized: wouldn't it be better to immediately convert strings to `&[u8]` **immediately** after normalization?

This would be like a marker for us, anything that's `&str` or similar is potentially not normalized, but as soon as we are done we just convert to bytes and forget about it.

With this change I guess you would make this function take a `&[u8]` directly, and do the conversion in the caller which as far as I can see is already normalizing correctly.

afilini commented on src/keys/bip39/pbkdf2.rs:
```diff
@@ -0,0 +1,87 @@
+/// Generate block (of given block_index) by calculating xor sum of iterations of PRF.
+fn xor_sum(hmac_prf: &HmacPRF, salt: &str, iter_count: u32, block_index: u32, block: &mut [u8]) {
+    // for the first iteration, we concat: salt + block_index (as big-endian bytes)
+    let mut prev_u = Vec::with_capacity(salt.len() + 4);
```
Since the length is fixed i guess this could also be an array with static length (there should be a constant in bitcoin_hashes for this).

I think the code would probably still look decently good with `copy_from_slice`: https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice

afilini reviewed (commented):
This is just a partial review, I still haven't looked at all the files.

I just wanted to post this comments so that you could start thinking about them and see if they make any sense.

evanlinjin commented:
> > Should we avoid using heap memory, and keep everything on the stack?
> 
> One advantage of this (on top of the extra safety) is that it would be much easier to then port to embedded hardware. We have many features in bdk which I guess are not really fit for hardware wallets, but mnemonics are for sure one that we'll need to have.
> 
> Are you able to do a rough estimation of how much longer/how much harder it would be to implement in this way?

Less than a week. But I'm stuck into multi descriptor wallet business 😅😂

afilini commented:
Yes, multi-descriptor is definitely the priority right now. We'll get back to this once you are done there :)

vladimirfomene commented on .gitignore:
```diff
@@ -3,3 +3,6 @@ Cargo.lock
 *.swp
 .idea
+
+# IDE
```
Nits: `.idea` is for IntelliJ. I don't know if it is necessary to have that `IDE` comment.

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,975 @@
+        // parse word indexes and ENT+CS bits from mnemonic words
+        let parse_result = sentence_words
+            .iter()
+            .map(|&word| word_to_index_map.get(word).unwrap_or(&utils::U11_EOF))
```
Why not throw an invalid word error here if the word is not in the HashMap? What is the utility of having `&utils::U11_EOF` as the default value here?

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,975 @@
+        }
+
+        let word_to_index_map = language.word_map();
+        let mut word_indexes = Vec::with_capacity(MS_MAX); // word indexes
+        let mut ent_cs_bits = Vec::with_capacity(MS_MAX * utils::U11_BITS); // ENT+CS bits
```
why not just use `ms` as your vector size instead of `MS_MAX`?

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,975 @@
+            .iter()
+            .map(|&word| word_to_index_map.get(word).unwrap_or(&utils::U11_EOF))
+            .try_for_each(|word_index| {
+                if *word_index > utils::U11_MAX {
```
Given that you are getting the `word_index` from the `word_map` is there a scenario where the `word_index` will be greater than `utils::U11_MAX`. I'm thinking if you throw an error for invalid words there will be no need for this if/else logic.

vladimirfomene commented on examples/rpcwallet.rs:
```diff
@@ -217,13 +218,10 @@ fn main() -> Result<(), Box<dyn Error>> {
-    // `Ok(mnemonic)` would also work if there's no passphrase and it would
-    // yield the same result as this construct with `password` = `None`.
-    Ok((mnemonic, password))
+    let mnemonic_with_passphrase: GeneratedKey<_, _> =
+        MnemonicWithPassphrase::generate((WordCount::Words12, Language::English, password))?;
+    Ok(mnemonic_with_passphrase)
```
Given that there is a change in the examples, I believe this will affect users. Is it possible to implement the BIP in such a way that it doesn't change anything for users.

vladimirfomene commented on src/keys/bip39/pbkdf2.rs:
```diff
@@ -0,0 +1,87 @@
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
```
It will be great to have a reference to the PBKDF2 RFC link as part of this module's documentation. https://datatracker.ietf.org/doc/html/rfc2898

vladimirfomene commented on src/keys/bip39/wordlists/mod.rs:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    /// Generate word map for given language.
+    pub fn word_map(&self) -> HashMap<&str, u16> {
```
can we write a test for this method?

evanlinjin replied:
What kind of test are you suggesting?

vladimirfomene replied:
I was thinking of writing a test to make sure we have correct word to indices mapping in the hashmap.

vladimirfomene reviewed (commented):
Thanks for helping us move this forward! Just a couple of questions and comments.

rajarshimaitra commented:
Is this a good idea to have it in bdk_core eventually? Or we wanna do key generation outside of core separately?

danielabrozzoni commented:
We closed #561, let's close this one as well :)
